python_modules/dagster/dagster/_core/instance/config.py

```diff
@@ -600,3 +603,74 @@ def dagster_instance_config_schema() -> Mapping[str, Field]:
     }
+
+
+class PoolGranularity(Enum):
+    OP = "op"
+    RUN = "run"
+
+
+@record
+class PoolConfig:
+    pool_granularity: Optional[PoolGranularity]
+    default_pool_limit: Optional[int]
+    op_granularity_run_buffer: Optional[int]
+
+
+@record
+class ConcurrencyConfig:
+    pool_config: PoolConfig
+    run_queue_config: Optional["RunQueueConfig"]
+
+    @staticmethod
+    def from_concurrency_settings(
+        concurrency_settings: Mapping[str, Any],
+        run_coordinator_run_queue_config: Optional["RunQueueConfig"],
+    ) -> "ConcurrencyConfig":
+        from dagster._core.run_coordinator.queued_run_coordinator import RunQueueConfig
+
+        run_settings = concurrency_settings.get("runs", {})
+        pool_settings = concurrency_settings.get("pools", {})
+        pool_granularity_str = pool_settings.get("granularity")
+        if run_coordinator_run_queue_config:
+            run_queue_config = RunQueueConfig(
+                max_concurrent_runs=run_settings.get(
+                    "max_concurrent_runs", run_coordinator_run_queue_config.max_concurrent_runs
+                ),
+                tag_concurrency_limits=run_settings.get(
+                    "tag_concurrency_limits",
+                    run_coordinator_run_queue_config.tag_concurrency_limits,
+                ),
+                max_user_code_failure_retries=run_coordinator_run_queue_config.max_user_code_failure_retries,
+                user_code_failure_retry_delay=run_coordinator_run_queue_config.user_code_failure_retry_delay,
+                should_block_op_concurrency_limited_runs=bool(pool_settings)
+                or run_coordinator_run_queue_config.should_block_op_concurrency_limited_runs,
+                op_concurrency_slot_buffer=pool_settings.get(
+                    "op_granularity_run_buffer",
+                    run_coordinator_run_queue_config.op_concurrency_slot_buffer,
+                ),
+            )
+        else:
+            run_queue_config = None
+
+        pool_granularity = PoolGranularity(pool_granularity_str) if pool_granularity_str else None
+        if (
+            not pool_granularity
+            and run_coordinator_run_queue_config
+            and run_coordinator_run_queue_config.should_block_op_concurrency_limited_runs
+        ):
+            # if this was explicitly configured in the run coordinator config, we should default to op granularity
+            pool_granularity = PoolGranularity.OP
+
+        default_limit = pool_settings.get("default_limit")
+        if default_limit is None:
+            default_limit = concurrency_settings.get("default_op_concurrency_limit")
+
+        return ConcurrencyConfig(
+            pool_config=PoolConfig(
+                pool_granularity=pool_granularity,
+                default_pool_limit=default_limit,
+                op_granularity_run_buffer=pool_settings.get("op_granularity_run_buffer"),
+            ),
+            run_queue_config=run_queue_config,
+        )
```

Comment:
just sanity checking here - after this you will still be able to impose a maximum number of runs in addition to limits at the op level? i think it is useful to have that as a stopgap

Reply:
You'll still be fully able to configure the run queue as before, even if the pool configuration is set to op granularity.